[DTensor] add op support: aten.squeeze_.dim

XilunWu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #159532

**Summary**
This PR enables in-place op `aten.squeeze_.dim` on DTensor with a change to
DTensor dispatch logic: when processing in-place operator, we should assign
`output_sharding.output_spec` back to the first argument. This is because
the in-place op_call on `arg._local_tensor` could also shift the tensor meta.

**Test**
`pytest test/distributed/tensor/test_view_ops.py -s -k  test_squeeze_`

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @tianyu-l

XilunWu committed:
[DTensor] add op support: aten.squeeze_.dim

[ghstack-poisoned]

XilunWu commented on torch/distributed/tensor/_dispatch.py:
```diff
@@ -265,6 +265,14 @@ def default_tensor(spec: DTensorSpec) -> torch.Tensor:
+                output_spec = output_sharding.output_spec
+                assert isinstance(output_spec, DTensorSpec)
+                assert isinstance(args[0], DTensor)
+                args[0]._spec = output_spec
```
@tianyu-l pointed out that, besides `DTensor._spec.tensor_meta`, the subclass' metadata also needs override.

XilunWu replied:
Edward suggests try `return_and_correct_aliasing` to fix the outer tensor meta, will try.

ezyang replied:
Is merged, rebase please

zpcore commented on torch/distributed/tensor/_ops/_view_ops.py:
```diff
@@ -677,6 +677,9 @@ def reshape_strategy(op_schema: OpSchema) -> StrategyType:
 
 register_op_strategy_map(aten.squeeze.default, torch.squeeze)
+register_op_strategy_map(
+    aten.squeeze_.dim, torch.squeeze, schema_info=RuntimeSchemaInfo(0)
```
Looks like the `RuntimeSchemaInfo(0)` will not take impact since the first arg is the self tensor. It will be hashed anyway: https://github.com/pytorch/pytorch/blob/78d7f0cdec9f87eb26f03ea38e2e5973de1d7964/torch/distributed/tensor/_op_schema.py#L456

XilunWu replied:
yeah it's a typo, should be `1`. Don't understand why I put `0` here lol.

zpcore commented on test/distributed/tensor/test_view_ops.py:
```diff
@@ -614,6 +616,18 @@ def test_view_redistribution(self):
 
+    @with_comms
+    def test_squeeze_(self):
+        mesh_2d = init_device_mesh(self.device_type, (3, 2), mesh_dim_names=("a", "b"))
```
Is the global size 6 set by intention?

XilunWu replied:
no, just to follow the existing tests. It's 6 because of those tests I guess

zpcore commented on test/distributed/tensor/test_view_ops.py:
```diff
@@ -614,6 +616,18 @@ def test_view_redistribution(self):
+        x = torch.randn((1, 4), device=self.device_type)
+        dist_x = DTensor.from_local(x, mesh_2d, [Partial(), Shard(1)])
+        self._test_op_on_dtensor(
+            torch.ops.aten.squeeze_.dim,
```
Should we also check if dist_x is changed or not?

XilunWu committed:
Update on "[DTensor] add op support: aten.squeeze_.dim"


**Summary**
This PR enables in-place op `aten.squeeze_.dim` on DTensor with a change to
DTensor dispatch logic: when processing in-place operator, we should assign
`output_sharding.output_spec` back to the first argument. This is because
the in-place op_call on `arg._local_tensor` could also shift the tensor meta.

**Test**
`pytest test/distributed/tensor/test_view_ops.py -s -k  test_squeeze_`

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta tianyu-l

[ghstack-poisoned]

AaronWang04 commented:
@XilunWu this is the table dump of the forward pass. as you said yeah it is missing a collective and all the values are off :/

main branch
```
Global
  FORWARD PASS
    *c10d_functional.all_reduce: 2
    **aten.view.default
      shape: [torch.Size([20, 5, 10])]
      sharding: [(Shard(dim=0),)]
      device mesh: DeviceMesh((4,), device: 'cuda', stride: (1,))
    **aten.view.default
    **aten.sum.default
      shape: [torch.Size([20, 5, 10])]
      sharding: [(Shard(dim=0),)]
      device mesh: DeviceMesh((4,), device: 'cuda', stride: (1,))
    **aten.sum.default
    **aten.ones_like.default
      shape: [torch.Size([])]
      sharding: [(Partial(sum),)]
      device mesh: DeviceMesh((4,), device: 'cuda', stride: (1,))
    **aten.ones_like.default
```


this PR
```
Global
  FORWARD PASS
    *c10d_functional.all_reduce: 1
    **aten.view.default
      shape: [torch.Size([20, 5, 10])]
      sharding: [(Shard(dim=0),)]
      device mesh: DeviceMesh((4,), device: 'cuda', stride: (1,))
    **aten.view.default
    **aten.sum.default
      shape: [torch.Size([20, 5, 10])]
      sharding: [(Shard(dim=0),)]
      device mesh: DeviceMesh((4,), device: 'cuda', stride: (1,))
    **aten.sum.default
    **aten.ones_like.default
      shape: [torch.Size([])]
      sharding: [(Partial(sum),)]
      device mesh: DeviceMesh((4,), device: 'cuda', stride: (1,))
    **aten.ones_like.default
```

XilunWu committed:
Update on "[DTensor] add op support: aten.squeeze_.dim"


**Summary**
This PR enables in-place op `aten.squeeze_.dim` on DTensor with a change to
DTensor dispatch logic: when processing in-place operator, we should assign
`output_sharding.output_spec` back to the first argument. This is because
the in-place op_call on `arg._local_tensor` could also shift the tensor meta.

**Test**
`pytest test/distributed/tensor/test_view_ops.py -s -k  test_squeeze_`

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta tianyu-l

[ghstack-poisoned]

XilunWu committed:
Update on "[DTensor] add op support: aten.squeeze_.dim"


**Summary**
This PR enables in-place op `aten.squeeze_.dim` on DTensor with a change to
DTensor dispatch logic: when processing in-place operator, we should assign
`output_sharding.output_spec` back to the first argument. This is because
the in-place op_call on `arg._local_tensor` could also shift the tensor meta.

**Test**
`pytest test/distributed/tensor/test_view_ops.py -s -k  test_squeeze_`

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta tianyu-l

[ghstack-poisoned]

zpcore reviewed (approved):
LGTM!

XilunWu commented:
@pytorchbot merge
